### PR TITLE
NTU microphysics scheme

### PR DESCRIPTION
**Pull request by tzuchin12**
TYPE: new feature

KEYWORDS: NTU microphysics scheme

SOURCE: Tzu-Chin Tsai and Jen-Ping Chen (National Taiwan University)

DESCRIPTION OF CHANGES:
Implementation of the NTU microphysics scheme (mp_physics = 56). 

The NTU (National Taiwan University) scheme applies double moments for the liquid-phase and triple moments for the ice-phase hydrometeors together with the consideration for ice crystal shape and density variations. There are five major features to NTU scheme:  
1. Condensation nuclei (CN) and ice nuclei (IN) are tracked separately for the processes of cloud/rain activation and ice deposition-nucleation using predicted supersaturation;  
2. Applying the triple-moment (the zeroth, second, and third moments) closure method to describe the evolution of ice particle’s spectrum; 
3. Solid-phase hydrometeors’ classification (pristine ice, snow aggregate, rimed ice, and hailstone) is redefined according to their key formation mechanisms; 
4. Ice crystals’ shape and apparent density can evolve gradually according to the growth conditions; 
5. Fall speed of each moment for frozen particles depends on shape and density.  

Recommendations:
Coarse-grid simulations should be okay with the NTU scheme when a time splitting integration is applied; Occasionally, the typical time step of 6dx (grid length of x-direction in kilometers) might be reduced to 4dx or 3dx for resolving the supersaturation explicitly in fine-grid (spacing less than 2 km) simulations. 

Documentation:
For the detailed parameterizations in the NTU scheme, the liquid-based formulae of the bulk conversion rates are adopted from [1], and the technique for aerosol activation to rain/cloud is introduced in [2] and [3]. The triple-moment approach, representation of ice properties, mixed- and solid-based equations/kernels are illustrated in [4] and [5]. In addition to the aerosol background type (CCNTY), some advanced physical options of crystal shape (SHAPE), apparent density (RHOX), fall speed relation (VTX), and bulk method (AFAX) concerning each hydrometeor are described on top of the main module.

LIST OF MODIFIED FILES:
Registry/Registry.EM_COMMON
Registry/Registry.NMM
Registry/registry.var
Registry/registry.wrfplus
dyn_em/solve_em.F
dyn_em/start_em.F
phys/Makefile
phys/module_microphysics_driver.F
phys/module_mp_ntu.F
phys/module_physics_init.F
share/module_bc.F
wrftladj/module_bc_ad.F
wrftladj/module_bc_tl.F
wrftladj/module_microphysics_driver_ad.F
wrftladj/module_microphysics_driver_tl.F
wrftladj/solve_em_ad.F
wrftladj/solve_em_tl.F
run/README.namelist

TESTS CONDUCTED: 
1. jenkins regression tests pass
2. 24-hr summer and winter cases conducted
3. The figure listed below is one real-case example of simulated surface rainfall using different microphysics schemes and different physical treatments within the NTU scheme. 
 

image

image

4. The background aerosols (QDCN, QTCN, and QINI) are needed for the initiation of liquid droplets (activation) and ice particles (nucleation) in the NTU scheme. If we do not specify aerosol boundary conditions, this may cause these inertial aerosols to vanish as a result of advection after a long simulation time. The below figure is one example of a real-case simulation with the default boundary condition (left panels) of QDCN at different time periods. Currently, the zero value (in flow_dep_bdy) and constant of ccn_conc (in flow_dep_bdy_qnn) assigned for inflow in the boundary specified region are not suitable for the vertical damping of aerosols in the NTU scheme. The right panels depict the revised results with the modified flow_dep_bdy_ntu. 

image

RELEASE NOTE: 

The NTU (National Taiwan University) scheme applies double moments for the liquid-phase and triple moments for the ice-phase hydrometeors together with the consideration for ice crystal shape and density variations. There are five major features to NTU scheme:  condensation nuclei (CN) and ice nuclei (IN) are tracked separately for the processes of cloud/rain activation and ice deposition-nucleation using predicted supersaturation;  applying the triple-moment (the zeroth, second, and third moments) closure method to describe the evolution of ice particle’s spectrum; solid-phase hydrometeors’ classification (pristine ice, snow aggregate, rimed ice, and hailstone) is redefined according to their key formation mechanisms; ice crystals’ shape and apparent density can evolve gradually according to the growth conditions; and fall speed of each moment for frozen particles depends on shape and density.  

For the detailed parameterizations in the NTU scheme, the liquid-based formulae of the bulk conversion rates are adopted from [1], and the technique for aerosol activation to rain/cloud is introduced in [2] and [3]. The triple-moment approach, representation of ice properties, mixed- and solid-based equations/kernels are illustrated in [4] and [5]. In addition to the aerosol background type (CCNTY), some advanced physical options of crystal shape (SHAPE), apparent density (RHOX), fall speed relation (VTX), and bulk method (AFAX) concerning each hydrometeor are described on top of the main module. 

References: 
[1] Chen, J.-P. and S.-T. Liu, 2004: Physically based two-moment bulkwater parameterization for warm-cloud microphysics. Quart. J. Roy. Meteor. Soc., 130, 51–78, doi:10.1256/qj.03.41. 
[2] Cheng, C.-T., W.-C. Wang, and J.-P. Chen, 2007: A modeling study of aerosol impacts on cloud microphysics and radiative properties. Quart. J. Roy. Meteor. Soc., 133, 283–297, doi:10.1002/qj.25. 
[3] Cheng, C.-T., W.-C. Wang, and J.-P. Chen, 2010: Simulation of the effects of increasing cloud condensation nuclei on mixed-phase clouds and precipitation of a front system. Atmos. Res., 96, 461–476, doi:10.1016/j.atmosres.2010.02.005. 
[4] Chen, J.-P. and T.-C. Tsai, 2016: Triple-moment modal parameterization for the adaptive growth habit of pristine ice crystals. J. Atmos. Sci., 73, 2105-2122, doi:10.1175/JAS-D-15-0220.1.
[5] Tsai, T.-C. and J.-P. Chen, 2020: Multi-moment ice bulk microphysics scheme with consideration for particle shape and apparent density. Part I: Methodology and idealized simulation. J. Atmos. Sci., 77, 1821–1850, doi:10.1175/JAS-D-19-0125.1.

### CODE REVIEW

File: Registry/Registry.NMM
Comment (weiwangncar) [inline Registry/Registry.NMM:880]: @tzuchin12 These are probably needed for the NMM registry.

File: phys/Makefile
Comment (weiwangncar) [inline phys/Makefile:106]: @tzuchin12 Can you line up this?

Comment (tzuchin12) [reply phys/Makefile:105]: Yes, done.

File: Registry/Registry.NMM
Comment (tzuchin12) [reply Registry/Registry.NMM:880]: I can not catch the point. They are already listed in Registry.NMM.

Comment (weiwangncar) [reply Registry/Registry.NMM:880]: @tzuchin12 I meant new variables qdcn, qtcn, qccn, qrcn and qnin.

**Comment by weiwangncar** [comment]
@tzuchin12 Please add your option to run/README.namelist file. Also for the RELEASE NOTE part of the PR, please add a short description of the scheme, in addition to the references.

File: Registry/registry.var
Comment (liujake) [inline Registry/registry.var:270]: You must use an older version of code here. These three lines of namelist change need to be recovered to the latest 'develop' branch code.

**Comment by liujake** [comment]
@tzuchin12 Can you please provide more information/reason for WRFDA-related modifications?, i.e., file # 3, 4, 12-17. Are those modifications just for making WRFPlus still compile with this new MP scheme added? In addition to WTF, did you do any other WRFDA-related compilation and tests? I need to do some WRFDA regression test before approving this PR.

File: Registry/Registry.NMM
Comment (tzuchin12) [reply Registry/Registry.NMM:880]: These five new variables are required to NTU_init through the usage of module_state_description in module_mp_ntu.F.

File: Registry/registry.var
Comment (tzuchin12) [reply Registry/registry.var]: Address review comments.

**Comment by tzuchin12** [comment]
> @tzuchin12 Please add your option to run/README.namelist file. Also for the RELEASE NOTE part of the PR, please add a short description of the scheme, in addition to the references.

Modified and updated as suggested.

**Comment by tzuchin12** [comment]
> @tzuchin12 Please add your option to run/README.namelist file. Also for the RELEASE NOTE part of the PR, please add a short description of the scheme, in addition to the references.

@weiwangncar  OK, modified and updated as suggested.

**Comment by tzuchin12** [comment]
> @tzuchin12 Can you please provide more information/reason for WRFDA-related modifications?, i.e., file # 3, 4, 12-17. Are those modifications just for making WRFPlus still compile with this new MP scheme added? 
Yes. 
In addition to WTF, did you do any other WRFDA-related compilation and tests? I need to do some WRFDA regression test before approving this PR.
Only WRFDA-related compilation was checked.

File: Registry/Registry.NMM
Comment (weiwangncar) [reply Registry/Registry.NMM:880]: @tzuchin12 I don't think these are needed. For example, if you check a variable used by P3 scheme, vmi3d, you will see that this variable, together with some others unique to P3 are not declared in NMM registry.

Comment (tzuchin12) [reply Registry/Registry.NMM]: Some modifications were made as suggested.

**Comment by liujake** [comment]
@tzuchin12 WRFDA regression tests passed. It is good to go for me. Very appreciate for your extra effort to take care of wrfplus-related changes.

File: Registry/Registry.EM_COMMON
Comment (davegill) [inline Registry/Registry.EM_COMMON:2894]: @tzuchin12 
Especially for the package declarations, please remove the 
```
#---------for ntu3m---
```
lines. They are OK for use in the namelist section.

File: Registry/Registry.NMM
Comment (davegill) [inline Registry/Registry.NMM:1777]: @tzuchin12 
Similarly here, you can remove the 
```
#---------for ntu3m---
```
lines for the package statements

File: Registry/registry.var
Comment (davegill) [inline Registry/registry.var:572]: @tzuchin12 
Zap the 
```
#---------for ntu3m---
``` 
here also

File: Registry/registry.wrfplus
Comment (davegill) [inline Registry/registry.wrfplus:887]: @tzuchin12 
Last ones here to remove
```
#---------for ntu3m---
```

File: dyn_em/start_em.F
Comment (davegill) [inline dyn_em/start_em.F:151]: @tzuchin12 
Should there be an IF test around this allocate statement? Is it required if the NTU scheme is not run?

File: dyn_em/start_em.F
Comment (davegill) [inline dyn_em/start_em.F:872]: @tzuchin12 
Here also, and I am assuming below. Are these required if NTU is not the MP chosen?

File: phys/Makefile
Comment (davegill) [reply phys/Makefile:105]: @tzuchin12 
This does not seem to be addressed yet. These "line up" requests are tricky. Usually, you need to pull the source code down, and fix it in an editor. The web-based approach has difficulty between distinguishing spaces or tabs.

File: phys/module_physics_init.F
Comment (davegill) [inline phys/module_physics_init.F:52]: @tzuchin12 
Break this into two lines to allow the conventional spacing at the end of the line

File: phys/module_physics_init.F
Comment (davegill) [inline phys/module_physics_init.F:1528]: @tzuchin12 
Most developers put a comment on a single line after the "&" character. So remove the 
```
!-------------------for ntu3m--
```
where you do not see others doing this.

File: phys/module_physics_init.F
Comment (davegill) [inline phys/module_physics_init.F:1530]: @tzuchin12 
Two lines to allow conventional "&" ending

File: share/module_bc.F
Comment (davegill) [inline share/module_bc.F:2618]: @tzuchin12 
Can you briefly describe how this routine is different from the other flow dependent boundary conditions?

Comment (davegill) [reply share/module_bc.F:2618]: Similarly for the a_ and tladj version

Comment (davegill) [reply share/module_bc.F:2618]: Sorry, the a_ and the g_versions also. How are these routines different than the existing flow dependent boundary routines?

File: dyn_em/solve_em.F
Comment (davegill) [inline dyn_em/solve_em.F:3768]: @tzuchin12 
Typically, developers denote their variables with a comment after the "&" character, instead of 
```
!----------------------------------for ntu3m----
```

**Comment by davegill** [comment]
@tzuchin12 @weiwangncar @kkeene44 
I have only a few very minor suggestions, regarding indenting, spacing, and comment conventions.

It might be appropriate to show figures of the same case, with a few different (similar) MP schemes. Maybe sensitivities to show the impact of the aerosol type?

Are there restrictions based on time step or grid distance, other suggestions for the users?

Please send a paragraph or two to Kelly for inclusion in the WRF User's Guide. Take a look at https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.2/users_guide_chap5.pdf at the "Summary of Microphysics Options" for a guideline. This is not urgent, but before the end of February.

File: Registry/Registry.EM_COMMON
Comment (tzuchin12) [reply Registry/Registry.EM_COMMON]: Modified as suggested.

File: Registry/Registry.NMM
Comment (tzuchin12) [reply Registry/Registry.NMM]: Modified as suggested.

File: Registry/registry.var
Comment (tzuchin12) [reply Registry/registry.var]: Modified as suggested.

File: Registry/registry.wrfplus
Comment (tzuchin12) [reply Registry/registry.wrfplus]: Modified as suggested.

File: dyn_em/start_em.F
Comment (tzuchin12) [reply dyn_em/start_em.F:151]: These two parameters are required for the NTU scheme only.

File: dyn_em/start_em.F
Comment (tzuchin12) [reply dyn_em/start_em.F:872]: They are required only for the NTU scheme.

File: phys/module_physics_init.F
Comment (tzuchin12) [reply phys/module_physics_init.F]: Modified as suggested.

File: phys/module_physics_init.F
Comment (tzuchin12) [reply phys/module_physics_init.F]: Modified as suggested.

File: phys/module_physics_init.F
Comment (tzuchin12) [reply phys/module_physics_init.F]: Modified as suggested.

File: share/module_bc.F
Comment (tzuchin12) [reply share/module_bc.F:2618]: The difference between the flow_dep_bdy_ntu and the other two existing routines is the allocated value to the field. The zero in flow_dep_bdy and one constant of ccn_conc in flow_dep_bdy_qnn are not suitable to be applied in the NTU scheme.

File: dyn_em/solve_em.F
Comment (tzuchin12) [reply dyn_em/solve_em.F]: Modified as suggested.

**Comment by davegill** [comment]
Jenkins
```
Please find result of the WRF regression test cases in the attachment. This build is for Commit ID: 8dea1410c50054bb0ce6ca548b6e721f5ee2ed3b, requested by: tzuchin12 for PR: https://github.com/wrf-model/WRF/pull/1364. For any query please send e-mail to David Gill.

    Test Type              | Expected  | Received |  Failed
    = = = = = = = = = = = = = = = = = = = = = = = =  = = = =
    Number of Tests        : 19           18
    Number of Builds       : 48           46
    Number of Simulations  : 163           161        0
    Number of Comparisons  : 103           102        0

    Failed Simulations are: 
    None
    Which comparisons are not bit-for-bit: 
    None
```

File: phys/Makefile
Comment (tzuchin12) [reply phys/Makefile]: Sorry, it was my misunderstanding. The correction has been made.

File: wrftladj/solve_em_ad.F
Comment (davegill) [inline wrftladj/solve_em_ad.F:3018]: @tzuchin12 
It looks like this code is purposefully commented out. Place your "!ntu3m" identifier AFTER the Fortran statement. Also at line 3039.

File: wrftladj/solve_em_tl.F
Comment (davegill) [inline wrftladj/solve_em_tl.F:2785]: @tzuchin12 
Another example to place the comment AFTER the Fortran statement (even though the Fortran statement is commented out).

File: wrftladj/solve_em_tl.F
Comment (davegill) [inline wrftladj/solve_em_tl.F:2805]: ... and here, too

**Comment by davegill** [comment]
@tzuchin12 
A GREAT JOB putting the comments into a standard form.

**Comment by davegill** [comment]
@tzuchin12 @kkeene44 
Here are outstanding issues to address. They were mentioned above, but you have handled some of the original list, and these are remaining:

It is appropriate to show figures of the same case, with a few different (similar) MP schemes. In your case, you could include sensitivities to show the impact of the aerosol options.

Are there restrictions based on time step or grid distance, other suggestions for the users? Are there cases that this works particularly well?

Please send a paragraph or two to Kelly for inclusion in the WRF User's Guide. Take a look at https://www2.mmm.ucar.edu/wrf/users/docs/user_guide_v4/v4.2/users_guide_chap5.pdf at the "Summary of Microphysics Options" for a guideline. This is not urgent, but before the end of February.

**Comment by weiwangncar** [comment]
@tzuchin12 @davegill Regarding the text in the UG, Tzu-Chin will provide me the paragraph and I will pass it to Kelly. We have also discussed preparing some simulation results to present here.

File: wrftladj/solve_em_ad.F
Comment (tzuchin12) [reply wrftladj/solve_em_ad.F]: Corrections have been made.

File: wrftladj/solve_em_tl.F
Comment (tzuchin12) [reply wrftladj/solve_em_tl.F]: Done.

File: wrftladj/solve_em_tl.F
Comment (tzuchin12) [reply wrftladj/solve_em_tl.F]: Done.

**Comment by tzuchin12** [comment]
@davegill @weiwangncar  The introduction to the NTU scheme in PR is updated, as well as the paragraph. Please let me know if anything I can do.

**Comment by davegill** [comment]
@tzuchin12 @weiwangncar @dudhia
This PR is shaping up very nicely. I think that all of the specific questions have been addressed. From a software perspective, this PR is ready.

**Comment by smileMchen** [comment]
@davegill @weiwangncar @dudhia @tzuchin12 
All,
Before you approve this PR and merge to repository, I would like to run a few real-data cases just to make sure it works fine. I will keep you updated after I am done. Thanks.

**Review by smileMchen — APPROVED** [review]

File: share/module_bc.F
Comment (dudhia) [inline share/module_bc.F:2623]: I would like to know why this new boundary condition is needed and especially why changes are needed for periodic conditions. The Registry has the PERIOD halos that may address these.

Comment (tzuchin12) [reply share/module_bc.F:2623]: The background aerosols (QDCN, QTCN, and QINI) are needed for the initiation of liquid droplets (activation) and ice particles (nucleation) in the NTU scheme. If we do not specify aerosol boundary conditions, this may cause these inertial aerosols to vanish as a result of advection after a long simulation time. The below figure is one example of a real-case simulation with the default boundary condition (left panels) of QDCN at different time periods. Currently, the zero value (in flow_dep_bdy) and constant of ccn_conc (in flow_dep_bdy_qnn) assigned for inflow in the boundary specified region are not suitable for the vertical damping of aerosols in the NTU scheme. The right panels depict the revised results with the modified flow_dep_bdy_ntu. It will be appreciated if you have any suggestions. 

image

**Comment by dudhia** [comment]
Do the background values have a profile or fixed value?
flow_dep_bdy_qnn allows a fixed value input.

**Comment by tzuchin12** [comment]
> Do the background values have a profile or fixed value? flow_dep_bdy_qnn allows a fixed value input.
The different initial values of background aerosols are derived from ntu_init (not fixed values, and there are four typical CCN types to choose).  Also, the vertical distribution is decreased exponentially with a scale height (3.57 km for continental and 0.8 km for marine type) except for the lowest three-sigma levels or below 850 hPa. So, it is not suitable to simply assign a profile or fixed value in the NTU scheme using flow_dep_bdy_qnn.

**Comment by dudhia** [comment]
OK, this keeps the inflow value constant at the initial value.
We can rename the routine more generally like flow_dep_bdy_fixed_inflow

**Comment by tzuchin12** [comment]
> OK, this keeps the inflow value constant at the initial value. We can rename the routine more generally like flow_dep_bdy_fixed_inflow
R: Thanks for the suggestion. The name of flow_dep_bdy_ntu has been changed to flow_dep_bdy_fixed_inflow over the related modules.

**Review by dudhia — APPROVED** [review]
also approve

**Comment by twjuliano** [comment]
@tzuchin12 @weiwangncar @davegill @dudhia
I think that there may have been a bug introduced by this PR in start_em.F, specifically on these lines (872-873):

`PRES(i,k,j) = grid%p(i,k,j)+grid%pb(i,k,j)              ! ntu3m`
`RHOA(i,k,j) = 1./grid%alt(i,k,j)*(1.+moist(i,k,j,P_QV)) ! ntu3m`

The loop goes from k=kts,kte where kte=kpe. Each of the variables on the RHS is defined at mass points, hence there should probably be a separate loop for these variables defined from k=kts,kte-1. The code fails on the second line above for an idealized simulation, as it tries to divide by zero.

**Comment by weiwangncar** [comment]
@twjuliano Thanks for pointing this out. I agree another do loop may be needed here.

**Comment by dudhia** [comment]
Yes, I think that loop should go to kte-1. Good catch.

**Comment by dudhia** [comment]
z_at_q seems to be for mass levels so that formula using ph is off by half
a level.

**Comment by dudhia** [comment]
z_at_q was added for Thompson aerosols and is not part of the NTU change.

**Comment by weiwangncar** [comment]
@dudhia Since z_at_q uses ph, it should be just height at full levels, whether it is intended or not.

**Comment by dudhia** [comment]
I don't think it is intended. There is already z_at_w for full levels.

**Comment by weiwangncar** [comment]
@dudhia Then that z_at_q calculation needs to be changed, and it may affect results from Thompson scheme. I'm ok with the change, but do we need Greg's blessing?

**Comment by dudhia** [comment]
I think we need to test the effect. It seems to be only used to initialize
the surface aerosol source.

**Comment by dudhia** [comment]
@Greg Thompson ***@***.***> Check which hgt you need in
thompson_init for aerosols.
Currently these are full level values but the name is z_at_q in start_em
which
seems to be to distinguish from z_at_w that is the full-level one. Should
we just
use z_at_w? From thompson_init, it seems the full level ones make sense.

**Comment by gthompsnWRF** [comment]
I am pretty sure that I need `z_at_w` because I need the "thickness" of the first model layer that represents the depth of atmosphere where the thermodynamic variables are defined.  That way, I compute an "airmass" variable that is used in the aerosol-aware surface emissions scaling.

**Comment by dudhia** [comment]
Yes, the name z_at_q is confusing. We could leave it for now, however.

**Pull request merged**
